hazelcast/src/main/java/com/hazelcast/internal/util/collection/MPSCQueue.java

```diff
@@ -0,0 +1,319 @@
+/*
+ * Copyright (c) 2008-2016, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.internal.util.collection;
+
+import com.hazelcast.util.concurrent.IdleStrategy;
+
+import java.util.AbstractQueue;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.hazelcast.util.Preconditions.checkNotNull;
+import static com.hazelcast.util.QuickMath.nextPowerOfTwo;
+import static java.util.concurrent.locks.LockSupport.park;
+import static java.util.concurrent.locks.LockSupport.unpark;
+
+/**
+ * Multi producer single consumer queue. This queue has a configurable {@link IdleStrategy} so if there is nothing to take,
+ * the thread can idle and eventually can do the more expensive blocking. The blocking is especially a concern for the putting
+ * thread, because it needs to notify the blocked thread.
+ *
+ * This MPSCQueue is based on 2 stacks; so the items are put in a reverse order by the putting thread, and by the taking thread
+ * they are reversed in order again so that the original ordering is restored. Using this approach, if there are multiple items
+ * on the stack, the owning thread can take them all using a single cas. Once this is done, the owning thread can process them
+ * one by one and doesn't need to content with the putting threads; reducing contention.
+ *
+ * @param <E>
```

Comment:
This is from the original JavaDoc: `the type of elements held in this collection`